finetuner/tuner/pytorch/miner.py

```diff
@@ -42,3 +42,71 @@ def mine(
         idxes_anchor, idxes_pos, idxes_neg = np.where(triplets)
         return list(zip(idxes_anchor, idxes_pos, idxes_neg))
+
+
+class SiameseSessionMiner(BaseMiner):
+    def mine(
+        self, embeddings: AnyTensor, labels: List[Tuple[int, int]]
+    ) -> List[Tuple[int, int, int]]:
+        """Generate tuples from input embeddings and labels.
+
+        :param embeddings: embeddings from model, should be a list of Tensor objects.
+        :param labels: labels of each embeddings, consist of session id and label.
+          the labels are either -1 or 1.
+        :return: a pair of label indices and their label as tuple.
+        """
+        assert len(embeddings) == len(labels)
+        rv = []
+        labels_with_index = [item + (index,) for index, item in enumerate(labels)]
+        sorted_labels_with_index = sorted(labels_with_index, key=lambda x: x[0])
+        for session, group in groupby(sorted_labels_with_index, lambda x: x[0]):
+            _, session_labels, session_indices = zip(*group)
+            session_pairs = []
+            for left, right in combinations(enumerate(session_labels), 2):
+                if left[1] >= 0 and right[1] >= 0:
+                    # 0 represents for anchor, 1 for positive, they form positive pairs.
+                    session_pairs.append((left[0], right[0], 1))
+                elif (left[1] >= 0 and right[1] == -1) or (
+                    left[1] == -1 and right[1] >= 0
+                ):
+                    # one of the item is positive or anchor, another one is negative
+                    session_pairs.append((left[0], right[0], -1))
+                else:
+                    # both are negatives
+                    continue
+            for left_idx, right_idx, label in session_pairs:
+                rv.append(
+                    (session_indices[left_idx], session_indices[right_idx], label)
+                )
+        return rv
+
+
+class TripletSessionMiner(BaseMiner):
+    def mine(
+        self, embeddings: AnyTensor, labels: List[Tuple[int, int]]
+    ) -> List[Tuple[int, int, int]]:
+        """Generate triplets from input embeddings and labels.
+
+        :param embeddings: embeddings from model, should be a list of Tensor objects.
+        :param labels: labels of each embeddings, consist of session id and label, the labels of
+          anchor, positive, negative are 0, 1 and -1.
+        :return: triplet of label indices follows the order of anchor, positive and negative.
+        """
+        assert len(embeddings) == len(labels)
+        rv = []
+        labels_with_index = [item + (index,) for index, item in enumerate(labels)]
+        sorted_labels_with_index = sorted(labels_with_index, key=lambda x: x[0])
+        for session, group in groupby(sorted_labels_with_index, lambda x: x[0]):
+            _, session_labels, session_indices = zip(*group)
+            for left, middle, right in permutations(enumerate(session_labels), 3):
```

Comment:
I'm thinking we can make this more efficient. Currently most of permutations we loop over will get discarded.

We already know that only positive and anchors can be at first two places, and negative has to be at the last place. So what you could do is to:
- first, get a list of positive/anchor indices
- get a list of negative indices
- loop over permutations of size 2 for pos/anchor
- inside this loop, loop over negative indices
- add each such combination to result